src/lustre/lustreAstDependencies.ml

```diff
@@ -237,36 +434,666 @@ let rec  mk_decl_map: LA.declaration IMap.t -> LA.t -> (LA.declaration IMap.t) g
+    
+let check_node_equations: dependency_analysis_data -> Lib.position -> LA.node_decl -> LA.node_decl graph_result =
+  fun ad pos ((i, imported, params, ips, ops, locals, items, contract_opt) as ndecl)->
+  (if not imported then
+     analyze_circ_node_equations ad.nsummary [] items 
+   else R.ok())
+  >> match contract_opt with
+     | None -> R.ok ndecl
+     | Some c ->
+        sort_and_check_contract_eqns ad pos (contract_suffix ^ "$" ^ i, params, ips, ops, c) >>=
+          fun (_, _, _, _, c') -> R.ok (i, imported, params, ips, ops, locals, items, Some c')
+(** Check if node equations do not have circularity and also, if the node has a contract
+    sort the contract equations. *)
+    
+let rec generate_summaries: dependency_analysis_data -> LA.t -> dependency_analysis_data
+  = fun ad ->
+  function
+  | [] -> ad
+  | LA.FuncDecl (_, ndecl) :: decls ->
+     let ns = mk_fun_summary ad.nsummary ndecl in
+     generate_summaries {ad with nsummary = IMap.union (fun k v1 v2 -> Some v2) ad.nsummary ns} decls
+  | LA.NodeDecl (_, ndecl) :: decls ->
+     let ns = mk_node_summary ad.nsummary ndecl in
+     generate_summaries {ad with nsummary = IMap.union (fun k v1 v2 -> Some v2) ad.nsummary ns} decls
+  | LA.ContractNodeDecl (_, cdecl) :: decls ->
+     let cs = mk_contract_summary ad.csummary cdecl in
+     generate_summaries {ad with csummary = IMap.union (fun k v1 v2 -> Some v2) ad.csummary cs } decls
+  | _ :: decls -> generate_summaries ad decls
+(** This function generates the node summary and contract summary data
+    This function requires that the program does not have any forward references. *)
+
+let rec sort_and_check_equations: dependency_analysis_data -> LA.t -> LA.t graph_result = 
+  fun ad ->
+  function
+  | (LA.TypeDecl _ as d) :: ds
+    | (LA.ConstDecl _ as d):: ds
+    | (LA.NodeParamInst _ as d) :: ds ->
+     sort_and_check_equations ad ds >>= fun ds' -> R.ok (d :: ds')
+  | LA.FuncDecl (pos, ndecl) :: ds ->
+     check_node_equations ad pos ndecl >>= fun ndecl' ->
+     sort_and_check_equations ad ds >>= fun ds' ->
+     R.ok (LA.FuncDecl (pos, ndecl') :: ds')
+  | LA.NodeDecl (pos, ndecl) :: ds ->
+     check_node_equations ad pos ndecl >>= fun ndecl' ->
+     sort_and_check_equations ad ds >>= fun ds' ->
+     R.ok (LA.NodeDecl (pos, ndecl') :: ds')
+  | LA.ContractNodeDecl (pos, contract_body) :: ds ->
+     sort_and_check_contract_eqns ad pos contract_body >>= fun contract_body' ->
+     sort_and_check_equations ad ds >>= fun decls' ->
+     R.ok (LA.ContractNodeDecl (pos, contract_body') :: decls' )  
+  | [] -> R.ok ([])
+(** sort equations for contracts and check if node and function equations have circular dependencies  *)
+        
+let sort_and_check_declarations decls =
+  (* Step 1. Sort the declarations according in their dependency order
+     This rules out the cases where we have recursive node or contract definitions *)
+  sort_decls mk_decl_map mk_graph_decls decls >>= fun sorted_decls -> 
```

Comment:
Why is `sort_decls` a high-order function? I would call `mk_decl_map` and `mk_graph_decls` directly in `sort_decls`. Use high-order functions when necessary, but avoid them if you can use a simple first-order function.

Reply:
no longer a higher order function